### ISSUE TITLE
Replacing FrozenBatchNorm with SyncBatchNorm?

### ISSUE THREAD

**Issue by kjgfcdb** [facebookresearch/maskrcnn-benchmark#561]
## 🚀 Feature
Recently pytorch-nightly has a new feature: [SyncBatchNorm](https://github.com/pytorch/pytorch/pull/14267)

I have tried to replace all `FrozenBatchNorm` in `maskrcnn_benchmark/modeling/backbone/resnet.py` with the new SyncBN, but I find that the program crashes after several iterations. Here is the log file:

```bash
2019-03-13 14:51:26,344 maskrcnn_benchmark.trainer INFO: Start training
2019-03-13 14:51:46,113 maskrcnn_benchmark.trainer INFO: eta: 1 day, 0:42:08  iter: 20  loss: 24.5855 (nan)  loss_box_reg: 0.0073 (nan)  loss_classifier: 13.8356 (nan)  loss_mask: 8.8279 (nan)  loss_objectness: 0.6801 (14225872939778.2500)  loss_rpn_box_reg: 0.1516 (3601105699730.5649)  time: 0.7603 (0.9883)  data: 0.0348 (0.0980)  lr: 0.007173  max mem: 6825
2019-03-13 14:52:01,234 maskrcnn_benchmark.trainer INFO: eta: 21:47:42  iter: 40  loss: nan (nan)  loss_box_reg: nan (nan)  loss_classifier: nan (nan)  loss_mask: nan (nan)  loss_objectness: 0.5964 (7112936469889.4238)  loss_rpn_box_reg: 0.1185 (1800552849865.3418)  time: 0.7537 (0.8722)  data: 0.0345 (0.0664)  lr: 0.007707  max mem: 6825
```

As you can see, there loss becomes `NaN` after several iterations.

I have also tried to use the normal `nn.BatchNorm2d`, and enlarge the batch size, but the problem didn't get solved neither.

So is it possible to use `SyncBatchNorm` here, in order to get a larger batch size?

By the way, I'm using 4 GPUS and I didn't change anything else.

**Comment by kjgfcdb** [comment]
My running command is 
```bash
export NGPUS=4
python -m torch.distributed.launch --nproc_per_node=$NGPUS tools/train_net.py --config-file "configs/e2e_mask_rcnn_R_50_FPN_1x.yaml"
```

**Comment by liu09114** [comment]
@lyakaap @fmassa
How to solve this problem?

**Comment by Iamal1** [comment]
@kjgfcdb sorry, would you please tell me how to use syncbn, I try to use the pytorch provided syncbn, and remove the broadcast=False in DDP, but some gpus will rise to 100% gpu usage, while others keep 0. I'm not familiar with DDP(set the rankid and process group, etc), could you give me some idea how to solve this, I'll appreciate it, thank you.

**Comment by kjgfcdb** [comment]
@Iamal1 In fact it is very easy to use syncbn, just simply replace all occurrences of `FrozenBatchNorm2d` in `resnet.py` with `nn. SyncBatchNorm`, then the program runs good. And you don't need to change anything else.

**Comment by Iamal1** [comment]
> @Iamal1 In fact it is very easy to use syncbn, just simply replace all occurrences of `FrozenBatchNorm2d` in `resnet.py` with `nn. SyncBatchNorm`, then the program runs good. And you don't need to change anything else.

That's what I've done,with a bit more modifications.
So what does the broadcast_buffer in the code means?
`    if distributed:
        model = torch.nn.parallel.DistributedDataParallel(
            model, device_ids=[local_rank], output_device=local_rank,
            # this should be removed if we update BatchNorm stats
            broadcast_buffers=False,
        )`
actually when I remove it, I also get similar problems using BN.

**Comment by kjgfcdb** [comment]
@Iamal1 Sorry for my misunderstanding. I was just told that there might be some problems in the details of the code. [This issue](https://github.com/facebookresearch/maskrcnn-benchmark/issues/267#issuecomment-489516019) may answer your question.

**Comment by zimenglan-sysu-512** [comment]
hi @kjgfcdb 
what's the performance of using SyncBn?

**Comment by kjgfcdb** [comment]
@zimenglan-sysu-512 Sorry, I made a mistake. The SyncBN still crashes in our program... So I'm also looking for solutions.

**Comment by zimenglan-sysu-512** [comment]
hi @kjgfcdb
have u tried sync-bn from apex?

**Comment by zhangliliang** [comment]
@kjgfcdb 

The crash problem might be caused by wrong weight initialization, i.e. loading the weight from `R-50.pkl`.
The moving mean and var has been merge in scale and bias in the weights of `R-50.pkl`.
When using `FrozenBatchNorm`, it is OK since its moving mean and var is 0 and 1.
But for `SyncBatchNorm` or `BatchNorm`, it would caluate the moving mean and var on each training batch, thus it would have some problem.

The solution might be straight-forward, that using `https://download.pytorch.org/models/resnet50-19c8e357.pth` for pretraining, instead of the `R-50.pkl`.

**Comment by kjgfcdb** [comment]
@zimenglan-sysu-512 Yes, I've tried that. But it crashed, too. Now I'am following zhangliliang's advice and will do some experiments about this problem.

**Comment by kjgfcdb** [comment]
@zhangliliang Thank you for your advice, I'll try this `pth` file.

**Comment by zhangliliang** [comment]
BTW, the .pth is from https://pytorch.org/docs/stable/torchvision/models.html, which is trained after normalizing the input to [0, 1] install of [0, 255]
Thus, it might be need to modified this config.
```
# Values to be used for image normalization
_C.INPUT.PIXEL_MEAN = [0.485, 0.456, 0.406]
# Values to be used for image normalization to
_C.INPUT.PIXEL_STD = [0.229, 0.224, 0.225]
# Convert image to BGR format (for Caffe2 models), in range 0-255
_C.INPUT.TO_BGR255 = False
```

**Comment by zimenglan-sysu-512** [comment]
@kjgfcdb 
how is it going?

**Comment by kjgfcdb** [comment]
@zimenglan-sysu-512 I've tried the pytorch pretrained model and set the `PIXEL_MEAN`, `PIXEL_STD` and `TO_BGR255` correctly. The results I got are as follows:
```bash
OrderedDict([('bbox', OrderedDict([('AP', 0.35498619907939205), ('AP50', 0.5646611776863764), ('AP75', 0.3837052968443872), ('APs', 0.1907768317757184), ('APm', 0.38030488740480484), ('APl', 0.47413005515848244)]))])
```
I run this experiment on 4 GPUs. The configuration file I used is `configs/e2e_faster_rcnn_R_50_FPN_1x.yaml`, and the only change I made on this file is to set `MODEL.BACKBONE.FREEZE_CONV_BODY_AT = -1`.
So it seems that my results can't match those without SyncBN, how about you?

**Comment by zimenglan-sysu-512** [comment]
@kjgfcdb 
i did what u has done, except that i train the model with mask-branch, and get `0.3651` for box and `0.333` for mask.

**Comment by kjgfcdb** [comment]
@zimenglan-sysu-512 Is it a typo? I think your results does not match the numbers listed in [Model Zoo](https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/MODEL_ZOO.md#end-to-end-faster-and-mask-r-cnn-baselines), which is not trained with SyncBN.

**Comment by zimenglan-sysu-512** [comment]
@kjgfcdb 
it is not a typo, i train `e2e_mask_rcnn_R_50_FPN_1x.yaml` using sync-bn, it's worse than not using it. so i have no idea how to fix it.

**Comment by amsword** [comment]
I have also got worse accuracy with SyncBN than FrozenBatchNorm. On my case, the input is fixed as 800x800. The baseline is 35%, and the accuracy with SyncBn is 33% without setting broadcast_buffer in DDP as True (or remove it).

**Comment by txytju** [comment]
Same thing happed to me. Performance dropped after using SyncBN instead of normal used BN.

**Comment by botcs** [comment]
Hey @kjgfcdb, @zimenglan-sysu-512 
based on the experimental results, this issue can be closed right?

**Comment by andics** [comment]
My experience with Mask-RCNN is that when the loss becomes NaN, the Learning rate must be decreased. There are several posts about this Loss issue in the repository.
